webpack: Transition app.js to be compiled by webpack.

armaanahluwalia opened the pull request:
This commit moves all files previously under the 'app' bundle in
the Django pipeline to being compiled by webpack under the 'app'
entry point. In the process, it moves assets under the app entry
to a file called app.js that consumes all relevant css and js files.

This commit also edits the webpack config to be able to expose certain
variables for third party libraries that are currently required by some
modules. This is bad coding form and should be refactored to requiring
whatever dependencies a module may have. This could be refactored in the
future. The variable exposure is done using expose-loader in webpack.

The app/index.html template is edited to override the newly introduced
'commonjs' block in the base template. This is done as a temporary
measure so as not to disrupt other pages on the app during the transition.

It also fixes the value of the 'this' context that was being inferred as
window by third party libraries. This is done using imports-loader in the
webpack config.

The last part of this commit edits all relevant js files to exposing their
local values to the window by setting 'window.var' = var; This is preferred
over expose-loader as using expose-loader for this may conflict with certain
loaders/plugins and also adds complexity to the build.

Thanks @priyankp10 for the initial work of exposing files to the window. That saved me a lot of time and in my opinion is the superior strategy to transition these files! :)

<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Transitioning app JS files to being compiled by webpack.

**Testing Plan:** <!-- How have you tested? -->
Tested app and other pages to make sure things are working.

**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented:
@armaanahluwalia you've got a few merged conflicts to fix.

Structurally, I'd prefer to keep the list of files in `webpack-assets.json` rather than having a separate `app.js`; is there a reason you moved that?

timabbott commented on tools/webpack.config.ts:
```diff
@@ -14,6 +14,34 @@ function getHotCSS(bundle:any[], isProd:boolean) {
+    ["../static/third/spectrum/spectrum.js", "this=>window"],
+    ["../static/third/jquery-throttle-debounce/jquery.ba-throttle-debounce.js", "this=>window"],
+    // AFAIK Lazyload is currently not being used and could be removed
+    ["../static/third/lazyload/lazyload.js", "this=>window"]
```
Indeed, I'll do a quick PR removing this.

armaanahluwalia commented:
@timabbott The reason I've moved asset import into app.js is because this is the way that webpack recommends one set-up a project. Each entry point must contain only one file. This helps webpack split out chunks and analyze the project as a whole.

The reason arrays exist on entry-points is for people developing libraries where files are peer dependent on each other. I'm not exactly sure what that use-case is myself but AFAIK that would be the only reason to use arrays as entry points.

Going forward setting up our files this way has the following advantages - 

1) Will not break or render webpack features useless. ( Main reason )
2) Allows for dynamic imports and code around imports.
3) Is easier to see what is happening since all definitions are inside the associated entry-point file itself.

armaanahluwalia commented:
The other main reason is that I plan on reusing the commons.js file as the entry file for 'commons'. This would not be possible AFAIK to import another entry point without copy and pasting all the files again.

timabbott commented:
OK.  Let's do this -- put any bundles of this form like `app.js` or `common.js` in `static/js/bundles/`, rather than the main `static/js`, which I think is  a pretty confusing place to put them.

priyankp10 commented:
Here what i currently have to say:
  *  The changes where `window.module = module` is added could be in it own commit.
  *  I think for the third, and node_module we should create a seperate bundle called `vendor` since most of the time it will not change and will cache better. 
  * Can you add a comment for what you doing in `webpack.config.ts` with the arrays, paths and expose and import loader?

priyankp10 commented on tools/webpack.config.ts:
```diff
@@ -14,6 +14,32 @@ function getHotCSS(bundle:any[], isProd:boolean) {
 }
+// Add variables within modules
+// Because of code corecing the value of window within the libraries
+var importsOptions = [
```
Use `const` here, and how about `importLoaderModules` for variable name.

priyankp10 commented on tools/webpack.config.ts:
```diff
@@ -14,6 +14,32 @@ function getHotCSS(bundle:any[], isProd:boolean) {
+// Expose Global variables for third party libraries to webpack modules
+// Use the unminified versions of jquery and underscore so that
+// Good error messages show up in production and development in the source maps
+var exposeOptions = [
```
Same here, use `const` and for variable name `exposeLoaderModules`.

priyankp10 commented on tools/webpack.config.ts:
```diff
@@ -131,6 +129,52 @@ export default (env?: string) : webpack.Configuration => {
+            nameArr = exposeArr.split('/');
+            nameArr = nameArr[nameArr.length - 1];
+            nameArr = nameArr.split('.');
+            name = nameArr[nameArr.length - 2];
```
This four lines of code if i interpreted correctly, is for getting `jquery` if the filename is `../../jquery.js`? If so this is what you should do.

```javascript
name = path.basename(path, '.js');
```

Couple of notes on it, you will need to import `path` module in full, i think currently it only imports resolve. And the variable `path` needs to be changed to `filePath` or `modulePath`.

priyankp10 commented on tools/webpack.config.ts:
```diff
@@ -131,6 +129,52 @@ export default (env?: string) : webpack.Configuration => {
+            use: useArr
+        });
+    }
+    config.module.rules = config.module.rules.concat(exposeLoaders);
```
I think `config.module.rules.push(...exposeLoaders)` is more clear and does the same job.

priyankp10 commented on tools/webpack.config.ts:
```diff
@@ -131,6 +129,52 @@ export default (env?: string) : webpack.Configuration => {
+            use: "imports-loader?" + importArgs
+        });
+    }
+    config.module.rules = config.module.rules.concat(importsLoaders);
```
Ditto.

armaanahluwalia commented:
Hey @timabbott I've made all the necessary changes to this PR.
@priyankp10 Thanks for your comments on this PR. They were really good suggestions :) I refactored a bunch of the code to be stricter and more easy to understand including some of the suggestions you had like pathname, maps etc.

Note 1: I have not split out chunks into a vendor bundle yet since the scope of this PR is to transition the app javascript bundles to webpack. I do plan on doing more intelligent code splitting once this is merged in and we move on to the babel step.
Note 2: In terms of splitting commits I split out the commits as suggested however some of the in between commits have broken tests. I wanted to get your opinion on what to squash but I think probably squashing the test fixes into the window global refactor makes sense.

Other than these two things we should be good to go. I'm not sure why these tests are failing it seems to be something to do with the provisioning but I don't feel like its related to any of the changes?

priyankp10 commented on tools/webpack-helpers.ts:
```diff
@@ -0,0 +1,72 @@
+   path: string;
+   args: string;
+}
+function getImportLoaders (optionsArr:Array<importLoaderOptions>) {
```
Mostly nit-picking, there is a extra space after function name, and it think it looks clean is so have space before the type annotation.

priyankp10 commented on tools/webpack-helpers.ts:
```diff
@@ -0,0 +1,72 @@
+   args: string;
+}
+function getImportLoaders (optionsArr:Array<importLoaderOptions>) {
+    let importsLoaders = [];
```
Use `const`.

priyankp10 commented on tools/webpack-helpers.ts:
```diff
@@ -0,0 +1,72 @@
+}
+function getImportLoaders (optionsArr:Array<importLoaderOptions>) {
+    let importsLoaders = [];
+    for(var loaderEntry of optionsArr) {
```
This code sorta use ES6, but then just there is a mix here are there about this ES5 stuff so can you use `const` here.

armaanahluwalia replied:
canst use const because were pushing into the array

priyankp10 replied:
What array are you talking about there? i meant `const loaderEntry` here. And pushing into a array which is declared `const` is not a error.

priyankp10 commented on tools/webpack-helpers.ts:
```diff
@@ -0,0 +1,72 @@
+    for(var loaderEntry of optionsArr) {
+        importsLoaders.push({
+            test: require.resolve(loaderEntry.path),
+            use: "imports-loader?" + loaderEntry.args
```
Use template string here `imports-loader?${loaderEntry.args}`

armaanahluwalia replied:
Not sure why the code needs to be more complicated than this.

priyankp10 replied:
As i said, this is mostly nit-picking, as i read through this is what popped up in my mind, feel free to ignore.

priyankp10 commented on tools/webpack-helpers.ts:
```diff
@@ -0,0 +1,72 @@
+}
+function getExposeLoaders (optionsArr:Array<exportLoaderOptions>) {
+    let exposeLoaders = [];
+    for(var loaderEntry of optionsArr) {
```
Ditto all the the comment i left above, use `const` and space after type declaration.

armaanahluwalia replied:
Again same reason.

priyankp10 commented on tools/webpack.config.ts:
```diff
@@ -1,6 +1,7 @@
 import { resolve } from 'path';
 import * as BundleTracker from 'webpack-bundle-tracker';
 import * as webpack from 'webpack';
+import {getExposeLoaders, getImportLoaders} from './webpack-helpers';
```
To be consistent add spaces here, `{ getExposeLoaders, getImportLoaders }`

armaanahluwalia commented:
@timabbott All conflicts resolved. Ready for review/merge.

armaanahluwalia commented:
@timabbott I may need access to SSH into the circle-ci build to debug this. I have reason to believe that the zulip-npm-cache is incorrect on circle-ci since things work for me locally. I think it was because for some reason the yarn.lock file had initially not been properly generated so imports-loader existed in package.json but somehow had not been included in the yarn.lock.

Could use your help in figuring out how else to debug this.

armaanahluwalia commented:
@timabbott I fixed the bug that was causing tests to fail. It was a timing bug because templates were being compiled after webpack on faster machines.

timabbott commented:
Makes sense; I just merged a2c9517f8dcd851b8d4fab475b90e096e57f2d1d  as a cleaner fix for that.  Note the commit message: We should be able to move this into some webpack configuration instead, and get rid of that process altogether.  I'll look at the rest soon.

timabbott commented:
@armaanahluwalia I pushed an updated version of this to your branch that fixes the webpack source map and double-hashing issues that I'd found before.  

The remaining problem is that it doesn't actually work in production; I get an exception the moment I load a page with not being able to find jquery.  Any ideas what's going on there?

I'll probably merge the early few commits soon, once I've tested them a bit more.

armaanahluwalia commented:
@timabbott Would like this to reviewed and merged in before a larger changeset is merged. There are no pending items on this PR. You mentioned attempting to use require statements instead of the window hack. I will open separate PRs for those changes.

armaanahluwalia commented:
@timabbott Rebased and updated this PR to work with latest changes.

timabbott commented:
Merged, thanks @armaanahluwalia!  I think the next step on this project (which I'd like you to start attempting) is to work on removing the `window` hack for one file in the `app` bundle.  Don't get deep into changing lots of files; I want to approve the approve the approach before we do it 100 times.
